### PR TITLE
Add copy/copied states to recovery phrase view

### LINKED ISSUE #400: Recovery Phrase copy button has no feedback that it worked
image

When clicking on the "Copy Recovery Phrase" button there is no feedback that copying has actually taken place. 

Standard practice on iOS seems to be to change the button text on tap to say "Copied" briefly or alternatively use a popover notification.

### PR DESCRIPTION
**Pull request by gordonbrander**
We make RecoveryPhraseView a component that knows how to manage its own state and side-effects, when stitched into a larger component.

Fixes #400.

Introduces:

- RecoveryPhraseModel
- RecoveryPhraseAction
- AppRecoveryPhraseCursor to map from app domain to recovery phrase domain

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 1]
```diff
@@ -63,6 +63,8 @@ struct AppView: View {
 
 //  MARK: Action
 enum AppAction: CustomLogStringConvertible {
+    case recoveryPhrase(RecoveryPhraseAction)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:66]: Wrapper action that boxes the RecoveryPhrase action in order to pass it down to the child component.

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 2]
```diff
@@ -134,6 +136,11 @@ enum AppAction: CustomLogStringConvertible {
     /// Set settings sheet presented?
     case presentSettingsSheet(_ isPresented: Bool)
 
+    /// Set recovery phrase on recovery phrase component
+    static func setRecoveryPhrase(_ phrase: String) -> AppAction {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:140]: This is a cheap trick for exposing child domain actions in parent context. It looks like a parent enum case, but it really is just a factory for a boxed child action.

File: xcode/Subconscious/Shared/Components/AppView.swift [hunk 3]
```diff
@@ -203,6 +226,12 @@ struct AppModel: ModelProtocol {
         environment: AppEnvironment
     ) -> Update<AppModel> {
         switch action {
+        case .recoveryPhrase(let action):
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/AppView.swift:229]: Handle boxed child actions by passing down to child component update function using update function synthesized by CursorProtocol.

File: xcode/Subconscious/Shared/Components/FirstRun/FirstRunCreateSphereView.swift [hunk 4]
```diff
@@ -17,11 +17,17 @@ struct FirstRunCreateSphereView: View {
                 VStack(alignment: .center, spacing: AppTheme.unit4) {
                     Text("Recovery Phrase")
                         .font(.headline)
-                    RecoveryPhraseView(text: app.state.sphereMnemonic ?? "")
+                    RecoveryPhraseView(
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/FirstRun/FirstRunCreateSphereView.swift:20]: Map state and send function from parent context to child context.

File: xcode/Subconscious/Shared/Components/FirstRun/FirstRunCreateSphereView.swift [hunk 5]
```diff
@@ -17,11 +17,17 @@ struct FirstRunCreateSphereView: View {
...
+                        send: Address.forward(
+                            send: app.send,
+                            tag: AppRecoveryPhraseCursor.tag
+                        )
+                    )
                     VStack(alignment: .leading, spacing: AppTheme.unit2) {
                         Text("This is your secret recovery phrase. You can use it to recover your data if you lose access.")
                             .foregroundColor(.secondary)
-                        Text("This is for your eyes only. We don't store it. Write it down. Keep it secret, keep it safe.")
+                        Text("This is for your eyes only. We don't store it. Write it down or add it to your password manager. Keep it secret, keep it safe.")
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/FirstRun/FirstRunCreateSphereView.swift:30]: Refining language a bit.

File: xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift [hunk 6]
```diff
@@ -42,10 +63,61 @@ struct RecoveryPhraseView: View {
...
+    ) -> Update<RecoveryPhraseModel> {
+        switch action {
+        case .setPhrase(let phrase):
+            var model = state
+            model.phrase = phrase
+            return Update(state: model)
+        case .copy:
+            // Copy to clipboard
+            UIPasteboard.general.string = state.phrase
+            logger.log("Copied recovery phrase to clipboard")
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift:93]: Log copies (useful for troubleshooting)

File: xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift [hunk 7]
```diff
@@ -42,10 +63,61 @@ struct RecoveryPhraseView: View {
     }
 }
 
+enum RecoveryPhraseAction: Hashable {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift:66]: Child component has its own actions, model, and update. It is required for the parent component to stitch these into its actions, model, and update.

File: xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift [hunk 8]
```diff
@@ -42,10 +63,61 @@ struct RecoveryPhraseView: View {
...
+        environment: Void
+    ) -> Update<RecoveryPhraseModel> {
+        switch action {
+        case .setPhrase(let phrase):
+            var model = state
+            model.phrase = phrase
+            return Update(state: model)
+        case .copy:
+            // Copy to clipboard
+            UIPasteboard.general.string = state.phrase
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift:92]: Is `UIPasteboard.general` always available? I'm imagining running this function in isolation as part of a test

**Review by bfollington — APPROVED** [review]
Nice, I'll use this as a reference

File: xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift (hunk 8 above)
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/FirstRun/RecoveryPhraseView.swift]: This was a good shout-out. Tests did work previously, but would copy things to your clipboard while running.

I have updated the PR to introduce a RecoveryPhraseEnvironment which includes a mockable pasteboard that defaults to UIPasteboard.general. I have also added unit tests.

**Review by bfollington — APPROVED** [review]
LGTM! Environment is a nice simple abstraction for cases like this one

**Pull request merged**